crates/volta-core/src/tool/package/fetch.rs

```diff
@@ -79,9 +76,90 @@ fn load_cached_distro(file: &Path, shasum_file: &Path) -> Option<Box<dyn Archive
-    debug!("Downloading {} from {}, to {}", &spec, &url, path.display());
-    Tarball::fetch(url, path).with_context(download_tool_error(spec, url.to_string()))
+fn fetch_remote_distro(
+    path: &Path,
+    name: &str,
+    details: &PackageDetails,
+    session: &mut Session,
+) -> Fallible<Box<Archive>> {
+    ensure_containing_dir_exists(&path).with_context(|_| ErrorDetails::ContainingDirError {
+        path: path.to_path_buf(),
+    })?;
+
+    let dir = path.parent().unwrap();
+
+    let command = npm_pack_command_for(name, &details.version.to_string()[..], session, Some(dir))?;
+    debug!("Running command: `{:?}`", command);
+
+    let spinner = progress_spinner(&format!(
+        "Downloading via {} npm pack to {}",
+        tool_version(name, details.version.to_string()),
+        dir.to_str().unwrap()
+    ));
+    let output = command.output()?;
+    spinner.finish_and_clear();
+
+    if !output.status.success() {
+        debug!(
+            "Command failed, stderr is:\n{}",
+            String::from_utf8_lossy(&output.stderr).to_string()
+        );
+        debug!("Exit code is {:?}", output.status.code());
+        throw!(ErrorDetails::NpmPackMetadataFetchError {
+            package: name.to_string(),
+        });
+    }
+
+    let response_json = String::from_utf8_lossy(&output.stdout);
+
+    let metadatas: Vec<super::serial::NpmPackData> = serde_json::de::from_str(&response_json)
+        .with_context(|_| ErrorDetails::NpmPackMetadataParseError {
+            package: name.to_string(),
+        })?;
+
+    debug!("Finding the first/only record from the output");
+    let metadata: super::serial::NpmPackData = match metadatas.iter().next() {
+        Some(data) => data.clone(),
+        // TODO: Make this be a correct error
+        None => throw!(ErrorDetails::PackageNotFound {
+            package: name.to_string()
+        }),
+    };
+
+    let tarball_from_npm_pack = dir.join(metadata.filename);
+
+    debug!("Moving the tarball to the expected path");
+    rename(tarball_from_npm_pack, path).with_context(|_| ErrorDetails::SetupToolImageError {
+        tool: name.into(),
+        version: details.version.to_string(),
+        dir: dir.to_path_buf(),
+    })?;
+
+    debug!("Attempting to load the now cached tarball from disk");
+    // TODO: Make this be a correct error
+    let distro = File::open(path).with_context(|_| ErrorDetails::PackageNotFound {
+        package: name.to_string(),
+    })?;
+
+    // TODO: Make this be a correct error
+    Tarball::load(distro).with_context(|_| ErrorDetails::PackageNotFound {
+        package: name.to_string(),
+    })
+}
+
+// build a command to run `npm pack` with json output
+fn npm_pack_command_for(
+    name: &str,
+    version: &str,
+    session: &mut Session,
+    current_dir: Option<&Path>,
+) -> Fallible<ToolCommand> {
+    let args = vec![
+        OsString::from("pack"),
+        OsString::from("--json"),
+        OsString::from(format!("{}@{}", name, version)),
```

Comment:
Since we only care about the file name, and not any of the other metadata that is output by `npm pack`, I think we can actually simplify this and use the standard, non-JSON output (which is already only the file name). We should also turn off the npm update notifier, since that's not relevant to our goal here:

```rust
let args = vec![
    OsString::from("pack"),
    OsString::from("--no-update-notifier"),
    OsString::from(format!("{}@{}", name, version)),
];
```

Then the `stdout` for the subcommand will be the filename